src/AltinnCore/Common/Services/Implementation/DataAppSI.cs

```diff
@@ -222,20 +227,40 @@ private static void ExtractAttachments(List<DataElement> dataList, List<Attachme
         }
 
         /// <inheritdoc />
-        public void DeleteFormAttachment(string org, string app, int instanceOwnerId, Guid instanceGuid, string attachmentType, string attachmentId)
+        public async Task<bool> DeleteFormAttachment(string org, string app, int instanceOwnerId, Guid instanceGuid, Guid dataGuid)
         {
             string instanceIdentifier = $"{instanceOwnerId}/{instanceGuid}";
-            List<AttachmentList> attachmentList = new List<AttachmentList>();
-            string apiUrl = $"instances/{instanceIdentifier}/data/{attachmentId}";
+            string apiUrl = $"instances/{instanceIdentifier}/data/{dataGuid}";
             string token = JwtTokenUtil.GetTokenFromContext(_httpContextAccessor.HttpContext, _cookieOptions.Cookie.Name);
             JwtTokenUtil.AddTokenToRequestHeader(_client, token);
 
-            Task<HttpResponseMessage> response = _client.DeleteAsync(apiUrl);
-            response.Result.EnsureSuccessStatusCode();
+            // Waiting for the instance and subsequently locking it.
+            SemaphoreSlim instanceLock = _instanceLocks.GetOrAdd(instanceGuid, k => new SemaphoreSlim(1, 1));
+            await instanceLock.WaitAsync();
+
+            try
+            {
+                HttpResponseMessage response = await _client.DeleteAsync(apiUrl);              
+
+                if (response.IsSuccessStatusCode)
+                {
+                    return true;
+                }
+                else
+                {
+                    _logger.LogError($"Deleting data element {dataGuid} for instance {instanceGuid} failed with status code {response.StatusCode}");
+                }
+            }
+            finally
+            {
+                instanceLock.Release();
+            }
+
+            return false;
         }
 
         /// <inheritdoc />
-        public async Task<Guid> SaveFormAttachment(string org, string app, int instanceOwnerId, Guid instanceGuid, string attachmentType, string attachmentName, HttpRequest attachment)
+        public async Task<DataElement> SaveFormAttachment(string org, string app, int instanceOwnerId, Guid instanceGuid, string attachmentType, string attachmentName, HttpRequest request)
```

Comment:
As we are looking into name changes, may be right time to rename the methods here aswell SaveFormAttachment can be renamed as SaveBinaryData for example.